src/unix/lwt_preemptive.ml

```diff
@@ -18,121 +18,71 @@ open Lwt.Infix
+  max_channel_queued := n
 
-(* The total number of preemptive threads currently running: *)
-let threads_count = ref 0
+(* The total number of domains currently running: *)
+let channels_count = ref 0
 
-(* +-----------------------------------------------------------------+
-   | Preemptive threads management                                   |
-   +-----------------------------------------------------------------+ *)
+  module C = Domainslib.Chan 
 
-module CELL :
-sig
-  type 'a t
-
-  val make : unit -> 'a t
-  val get : 'a t -> 'a
-  val set : 'a t -> 'a -> unit
-end =
-struct
-  type 'a t = {
-    m  : Mutex.t;
-    cv : Condition.t;
-    mutable cell : 'a option;
-  }
-
-  let make () = { m = Mutex.create (); cv = Condition.create (); cell = None }
-
-  let get t =
-    let rec await_value t =
-      match t.cell with
-      | None ->
-        Condition.wait t.cv t.m;
-        await_value t
-      | Some v ->
-        t.cell <- None;
-        Mutex.unlock t.m;
-        v
-    in
-    Mutex.lock t.m;
-    await_value t
-
-  let set t v =
-    Mutex.lock t.m;
-    t.cell <- Some v;
-    Mutex.unlock t.m;
-    Condition.signal t.cv
-end
-
-type thread = {
-  task_cell: (int * (unit -> unit)) CELL.t;
-  (* Channel used to communicate notification id and tasks to the
-     worker thread. *)
-
-  mutable thread : Thread.t;
-  (* The worker thread. *)
-
-  mutable reuse : bool;
-  (* Whether the thread must be re-added to the pool when the work is
-     done. *)
+
+type message = Do of (unit -> unit) | Quit 
+
+type chan = {
+  req : message C.t;
+  resp : unit C.t;
 }
 
-(* Pool of worker threads: *)
-let workers : thread Queue.t = Queue.create ()
+(* Pool of worker channels: *)
+let channels : chan Queue.t = Queue.create ()
 
 (* Queue of clients waiting for a worker to be available: *)
-let waiters : thread Lwt.u Lwt_sequence.t = Lwt_sequence.create ()
-
-(* Code executed by a worker: *)
-let rec worker_loop worker =
-  let id, task = CELL.get worker.task_cell in
-  task ();
-  (* If there is too much threads, exit. This can happen if the user
-     decreased the maximum: *)
-  if !threads_count > !max_threads then worker.reuse <- false;
-  (* Tell the main thread that work is done: *)
-  Lwt_unix.send_notification id;
-  if worker.reuse then worker_loop worker
-
-(* create a new worker: *)
-let make_worker () =
-  incr threads_count;
-  let worker = {
-    task_cell = CELL.make ();
-    thread = Thread.self ();
-    reuse = true;
+let waiters : chan Lwt.u Lwt_sequence.t = Lwt_sequence.create ()
+
+(* Code executed by a channel: *)
+let rec worker_loop channel =
+  match C.recv channel.req with
+  | Do f -> f (); C.send channel.resp (); worker_loop channel ()
+  | Quit -> ()
+
+(* create a new channel: *)
+let make_channel_and_domain () =
+  incr channels_count;
+  let channel = {
+    req : C.make 1;
+    resp : C.make 1;
   } in
-  worker.thread <- Thread.create worker_loop worker;
```

Comment:
something like this needs to present where a domain is assigned instead of a thread

Reply:
I'm not sure if this is the right way
```
  let worker = {
    req = C.make 1;
    resp = C.make 1;
    domain = Domain.spawn (fun () -> worker_loop worker);
    reuse = true;
  } in
  worker
```